[Security][Firewall] Passing the newly generated security token to the event during user switching

klandaika opened the pull request:
Event allows listeners to easily switch out the token if custom token updates are required

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Updated SwitchUserEvent to include the generated security Token. Allows the listeners to replace the token with their own (in case an application has some custom logic for token generation). The SwitchUserListener will now use the token returned by the event, so if token was not changed the self generated token will be used. If token was changed in the event then the new token would get used.

Reasons for this feature
--------------------------

In our current project users can have different Role sets depending on which organization they switch to. Our `User->getRoles()` always returns ["ROLE_USER"] and after login user is presented with choice of organizations they want to work in. Based on selected organization roles get updated with then stored token.  

Without the change proposed in this PR. The only way we can setup the proper roles during user switch is by replacing `security.authentication.switchuser_listener` service with our own implementation of the listener. 

With the proposed change, we can replace the security token with the one having all the roles we require directly inside our listener for `security.switch_user` event that gets thrown by Symfony's `SwitchUserListener`

nicolas-grekas commented on src/Symfony/Component/Security/Http/Event/SwitchUserEvent.php:
```diff
@@ -24,11 +25,13 @@ class SwitchUserEvent extends Event
+    private $token;
 
-    public function __construct(Request $request, UserInterface $targetUser)
+    public function __construct(Request $request, UserInterface $targetUser, TokenInterface $token)
```
adding a new required argument is a BC break. Instead, you should either make it optional, or rely solely on the setter

klandaika replied:
Got it. Made it optional.

klandaika commented:
Would you like me to do anything else in regards to this?

xabbuh commented on src/Symfony/Component/Security/Http/Event/SwitchUserEvent.php:
```diff
@@ -46,4 +49,20 @@ public function getTargetUser()
     }
+
+    /**
+     * @return TokenInterface
```
`TokenInterface|null`

xabbuh commented on src/Symfony/Component/Security/Http/Event/SwitchUserEvent.php:
```diff
@@ -46,4 +49,20 @@ public function getTargetUser()
+    }
+
+    /**
+     * @param TokenInterface $token
```
this docblock is not needed

xabbuh commented:
Can you also add a test that checks that the token can be replaced?

klandaika committed:
Passing the newly generated security token to the event during user switching.

Event allows listeners to easily switch out the token if custom token updates are required

klandaika commented:
Did all that was requested. Please check again.

nicolas-grekas commented:
Moving to milestone 3.4 as 3.3 is about to be closed.

xabbuh commented on src/Symfony/Component/Security/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 -----
 
+ * added `TokenInterface` to `\Symfony\Component\Security\Http\Event\SwitchUserEvent` to allow listeners to switch out 
+   the token when custom token generation is required by application.
```
This must now be done under a new `3.4.0` headline

klandaika replied:
Updated

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+        $this->tokenStorage->setToken($token);
+        $this->request->query->set('_switch_user', 'kuba');
+
+        $this->accessDecisionManager->expects($this->once())
```
`$this->any()` (it does not matter how often the method is called)

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+        $this->userProvider->expects($this->once())
+            ->method('loadUserByUsername')->with('kuba')
+            ->will($this->returnValue($user));
+        $this->userChecker->expects($this->once())
```
this doesn't look necessary

klandaika replied:
This was done in previous test, figured since accessDecision manager calls `checkPostAuth` we should have it mocked to expect to be called.

klandaika replied:
Removed, from test, but let me know if you want me to put it back.

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+            ->method('decide')->with($token, array('ROLE_ALLOWED_TO_SWITCH'))
+            ->will($this->returnValue(true));
+
+        $this->userProvider->expects($this->once())
```
`any()`

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+                }));
+
+        $listener = new SwitchUserListener($this->tokenStorage, $this->userProvider, $this->userChecker, 'provider123',
+            $this->accessDecisionManager, null, '_switch_user', 'ROLE_ALLOWED_TO_SWITCH', $dispatcher);
```
please do not wrap lines here

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+    {
+        $token = new UsernamePasswordToken('username', '', 'key', array('ROLE_FOO'));
+        $user = new User('username', 'password', array());
+        $replacedToken = new UsernamePasswordToken('replaced', '', 'key', array('ROLE_BAR'));
```
`key` should be `provider123` (or the other way around) to be consistent with the `SwitchUserListener` definition below

xabbuh replied:
instead of `replaced` I would pass the created `User` object as the first argument instead

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+
+    public function testSwitchUserWithReplacedToken()
+    {
+        $token = new UsernamePasswordToken('username', '', 'key', array('ROLE_FOO'));
```
`key` should be `provider123` (or the other way around) to be consistent with the `SwitchUserListener` definition below

xabbuh commented on src/Symfony/Component/Security/Http/Tests/Firewall/SwitchUserListenerTest.php:
```diff
@@ -227,4 +227,44 @@ public function testSwitchUserKeepsOtherQueryStringParameters()
+    public function testSwitchUserWithReplacedToken()
+    {
+        $token = new UsernamePasswordToken('username', '', 'key', array('ROLE_FOO'));
+        $user = new User('username', 'password', array());
```
instead of `username` I would pass the below created `User` object as the first argument instead

xabbuh commented on src/Symfony/Component/Security/CHANGELOG.md:
```diff
@@ -13,6 +13,8 @@ CHANGELOG
 3.4.0
 -----
 
+ * added `TokenInterface` to `\Symfony\Component\Security\Http\Event\SwitchUserEvent` to allow listeners to switch out 
```
I think it's more important to document that you can change the token being used by calling the `setToken()` method:

> * added a `setToken()` method to the `SwitchUserEvent` class to allow to replace the created token while switching users

chalasr commented:
Seems legit to me. @klandaika would you mind to rebase this PR against the 3.4 branch?

nicolas-grekas commented:
ping @chalasr @xabbuh for review

xabbuh commented on src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php:
```diff
@@ -145,8 +145,10 @@ private function attemptSwitchUser(Request $request)
-            $switchEvent = new SwitchUserEvent($request, $token->getUser());
+            $switchEvent = new SwitchUserEvent($request, $token->getUser(), $token);
             $this->dispatcher->dispatch(SecurityEvents::SWITCH_USER, $switchEvent);
+            //use the token from the event in case any listeners have replaced it.
```
missing space after `//`

klandaika replied:
Fixed

xabbuh commented on src/Symfony/Component/Security/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 3.4.0
 -----
 
+ * added a setToken() method to the SwitchUserEvent class to allow to replace the created token while switching users
```
the method and class names should be enclosed with backticks:

```markdown
 * added a `setToken()` method to the `SwitchUserEvent` class to allow to replace the created token while switching users
```

klandaika replied:
Fixed

chalasr commented:
Thank you @klandaika.
